torch/utils/_config_module.py

```diff
@@ -140,15 +141,18 @@ def __init__(self) -> None:
     def __setattr__(self, name: str, value: object) -> None:
         if name in self._bypass_keys:
             super().__setattr__(name, value)
-        elif name not in self._allowed_keys:
+        elif name not in self._config:
             raise AttributeError(f"{self.__name__}.{name} does not exist")
         else:
-            self._config[name] = value
+            self._config[name].user_override = value
             self._is_dirty = True
 
     def __getattr__(self, name: str) -> Any:
         try:
-            return self._config[name]
+            config = self._config[name]
+            if config.user_override is not None:
+                return copy.deepcopy(config.user_override)
+            return copy.deepcopy(config.default)
```

Comment:
I'm deeply suspicious about the deepcopy here. Do you really want to do it this way? There's some benefit but it just seems too magic, let's just make sure use sites don't mutate stuff?

Also, suppose someone says torch._dynamo.config.blah.append('foobar'). This is now BC breaking. Are you sure you wanted that? (Well, the real answer is to only allow immutable things in the config.....)

Reply:
Fair - Let me keep the existing reference semantics, and I can follow up with a PR if we want to change the behaviour here (I want to see the usage of lists and dicts in configs).